src/librustc_session/session.rs

```diff
@@ -1056,6 +1067,37 @@ fn build_session_(
         _ => CtfeBacktrace::Disabled,
     });
 
+    // Try to find a directory containing the Rust `src`, for more details see
+    // the doc comment on the `real_rust_source_base_dir` field.
+    let compute_real_rust_source_base_dir = || {
+        // A file we expect to find inside the `real_rust_source_base_dir`.
+        const SRC_FILE_TO_PROBE_FOR: &str = "src/libstd/lib.rs";
+
+        // First look in the location used by the `rust-src` `rustup` component.
+        let sysroot_candidate = sysroot.join("lib/rustlib/src/rust");
+        if sysroot_candidate.join(SRC_FILE_TO_PROBE_FOR).is_file() {
+            return Some(sysroot_candidate);
+        }
+
+        // As a fallback, check if the `sysroot` is inside a Rust build dir,
+        //  i.e. it ends in `build/$CFG_COMPILER_HOST_TRIPLE/stage{0-sysroot,1,2,3}`.
+        let mut sysroot_components = sysroot.components();
+        let last = sysroot_components.next_back()?.as_os_str();
+        if last == "stage0-sysroot" || last == "stage1" || last == "stage2" || last == "stage3" {
+            if sysroot_components.next_back()?.as_os_str() == host_triple {
+                if sysroot_components.next_back()?.as_os_str() == "build" {
```

Comment:
Note that on CI, at least, this will fail -- there we sometimes have the build directory in `obj/build`. Maybe we should have rustbuild tell the compiler what the prefix to strip is instead?

Reply:
We might want to do the symlink approach instead then, where `src/bootstrap` symlinks the checkout to `build/$triple/stage{0-sysroot,1,2,3}/lib/rustlib/src/rust`.

I just don't know what we'll have to do on platforms where that's not directly an option.

Reply:
Bonus: it would make `cargo`'s `-Z build-std` and presumably RLS, able to find the Rust sources, for locally built `rustup` toolchains.

Reply:
Hm, why would we want the symlink approach over an environment variable read or something (at runtime, I guess, set during our tests?)

Something like `RUST_SRC_PATH` which I think other tooling (RLS, perhaps?) uses

Reply:
But to be clear I think symlinking is fine (or hardlinking I guess, too, though could be more finicky).

I don't know whether platform support would be a problem -- if we can get it past bors though seems like we should land it, and we can make it optional I think, in the sense that you need it for tests to pass but if you don't have it things won't fail to build entirely.

Reply:
> Hm, why would we want the symlink approach over an environment variable read or something (at runtime, I guess, set during our tests?)

Because `rustup` places `rust-src` in a certain directory, and presumably distros can too, so we'd want to use that, and using a symlink would transparently let that work for local build sysroots, without adding an insta-stable env var.

> and we can make it optional I think, in the sense that you need it for tests to pass but if you don't have it things won't fail to build entirely.

Ah, good point, do we have a way for `src/bootstrap` to emit a warning if it tried something and that failed?
I'm thinking I would try to create the symlink, then do the same check `rustc` does (look for `src/libstd/lib.rs` inside there), and if either fails, warn instead of failing the build.

Reply:
Just eprintln should work fine.

I was suggesting that we use the environment variable as the fall back to normal detection, not as the only option. But I agree I think that if we can get the symlinks to work well we should.

Reply:
@eddyb would that symlink also supersede the [special treatment Miri currently needs](https://github.com/rust-lang/miri/blob/master/src/bin/cargo-miri.rs#L309-L329) to find the sources when built against a local toolchain?

Reply:
@RalfJung Oh wow I wasn't aware of that hack, I wonder if other tools also do this.
Yes, although I'm not sure what will happen on Windows.
*However*, [`std::os::windows::fs::symlink_dir`](https://doc.rust-lang.org/std/os/windows/fs/fn.symlink_dir.html) does exist, presumably it creates a NTFS "junction"?

So we might be set on all platforms we care about, barring any odd filesystems where `symlink`s aren't supported or banned by system policy.

Reply:
@eddyb it won't work for every Windows user though: https://github.com/rust-lang/rust/blob/30ddb5a8c1e85916da0acdc665d6a16535a12dd6/src/libstd/sys/windows/fs.rs#L760-L763

Reply:
@eddyb rustbuild already has a `symlink_dir` function which should do the right thing: https://github.com/rust-lang/rust/blob/99009bf31992fb9257b84550dbf7cb509f987952/src/bootstrap/util.rs#L131-L133

Reply:
Yeah I'll be using the rustbuild function. Also, this will only affect CI and a tiny subset of people building Rust (either using `remap-debuginfo = true` like CI, or testing `miri`, if `miri` removes it hack). And won't break the build if symlinking fails.

Reply:
Miri tested via `./x.py test src/tools/miri` [does not even rely on this hack](https://github.com/rust-lang/rust/blob/235938d1acdd93d6641a741c81f64e415b786751/src/bootstrap/test.rs#L388).